pkg/ratelimiter/rate_limiter.go

```diff
@@ -31,88 +31,98 @@ type Entry struct {
 	mutex sync.Mutex
 }
 
+// Limit controls token refilling for bucket entries
+type Limit struct {
+	// Maximum number of tokens that can be accumulated
+	MaxTokens uint16
+	// Number of tokens to refill per minute
+	RatePerMinute uint16
+}
+
+func (l Limit) Refill(entry *Entry) {
+	now := time.Now()
+	entry.mutex.Lock()
+	defer entry.mutex.Unlock()
+	minutesSinceLastSeen := now.Sub(entry.lastSeen).Minutes()
+	if minutesSinceLastSeen > 0 {
+		// Only update the lastSeen if it has been >= 1 minute
+		// This allows for continuously sending nodes to still get credits
+		entry.lastSeen = now
+		// Convert to ints so that we can check if above MAX_UINT_16
+		additionalTokens := int(l.RatePerMinute) * int(minutesSinceLastSeen)
+		// Avoid overflows of UINT16 when new balance is above limit
+		if additionalTokens+int(entry.tokens) > MAX_UINT_16 {
+			additionalTokens = MAX_UINT_16 - int(entry.tokens)
+		}
+		entry.tokens = minUint16(entry.tokens+uint16(additionalTokens), l.MaxTokens)
+	}
+}
+
 // TokenBucketRateLimiter implements the RateLimiter interface
 type TokenBucketRateLimiter struct {
-	log     *zap.Logger
-	wallets map[string]*Entry
-	mutex   sync.RWMutex
+	log           *zap.Logger
+	buckets       map[string]*Entry
+	mutex         sync.RWMutex
+	PriorityLimit Limit
+	RegularLimit  Limit
 }
 
 func NewTokenBucketRateLimiter(log *zap.Logger) *TokenBucketRateLimiter {
 	tb := new(TokenBucketRateLimiter)
 	tb.log = log.Named("ratelimiter")
-	tb.wallets = make(map[string]*Entry)
+	tb.buckets = make(map[string]*Entry)
```

Comment:
It occurs to me that we never clean up this map. We should at least make a todo to periodically purge expired entries, otherwise it'll slowly eat up all the system memory.